mavros_extras/src/plugins/wheel_odometry.cpp

```diff
@@ -0,0 +1,525 @@
+		twist_cov.twist.linear.z = 0.0;
+		// angular
+		twist_cov.twist.angular.x = 0.0;
+		twist_cov.twist.angular.y = 0.0;
+		twist_cov.twist.angular.z = rtwist(2);
+		// covariance
+		ftf::EigenMapCovariance6d twist_cov_map(twist_cov.covariance.data());
+		twist_cov_map.setZero();
+		twist_cov_map.block<3, 3>(0, 0).diagonal() << rtwist_cov(0), rtwist_cov(1), -1.0;
+		twist_cov_map.block<3, 3>(3, 3).diagonal() << -1.0, -1.0, rtwist_cov(2);
+
+		// Publish twist
+		if (twist_send) {
+			auto twist_cov_t = boost::make_shared<geometry_msgs::TwistWithCovarianceStamped>();
+			// header
+			twist_cov_t->header.stamp = time;
+			twist_cov_t->header.frame_id = frame_id;
+			// twist
+			twist_cov_t->twist = twist_cov;
+			// publish
+			twist_pub.publish(twist_cov_t);
+		}
+		// Publish odometry (only if we have initial yaw)
+		else if (yaw_initialized) {
+			auto odom = boost::make_shared<nav_msgs::Odometry>();
+			// header
+			odom->header.stamp = time;
+			odom->header.frame_id = frame_id;
+			odom->child_frame_id = child_frame_id;
+			// pose
+			odom->pose.pose.position.x = rpose(0);
+			odom->pose.pose.position.y = rpose(1);
+			odom->pose.pose.position.z = 0.0;
+			odom->pose.pose.orientation = quat;
+			ftf::EigenMapCovariance6d pose_cov_map(odom->pose.covariance.data());
+			pose_cov_map.setZero();
+			pose_cov_map.block<3, 3>(0, 0).diagonal() << rpose_cov(0), rpose_cov(1), -1.0;
+			pose_cov_map.block<3, 3>(3, 3).diagonal() << -1.0, -1.0, rpose_cov(2);
+			// twist
+			odom->twist = twist_cov;
+			// publish
+			odom_pub.publish(odom);
+		}
+
+		// Publish TF (only if we have initial yaw)
+		if (tf_send && yaw_initialized) {
+			geometry_msgs::TransformStamped transform;
+			// header
+			transform.header.stamp = time;
+			transform.header.frame_id = tf_frame_id;
+			transform.child_frame_id = tf_child_frame_id;
+			// translation
+			transform.transform.translation.x = rpose(0);
+			transform.transform.translation.y = rpose(1);
+			transform.transform.translation.z = 0.0;
+			// rotation
+			transform.transform.rotation = quat;
+			// publish
+			m_uas->tf2_broadcaster.sendTransform(transform);
+		}
+	}
+
+	/**
+	 * @brief Update odometry (currently, only 2-wheels differential configuration implemented).
+	 * Odometry is computed for robot's origin (IMU).
+	 * @param distance	distance traveled by each wheel since last odometry update
+	 * @param dt		time elapse since last odometry update (s)
+	 */
+	void update_odometry(std::vector<float> distance, float dt)
+	{
+		// Currently, only 2-wheels configuration implemented
+		int count = std::min(2, (int)distance.size());
+		switch (distance.size())
+		{
+		// Differential drive robot.
+		// The wheels are assumed to be parallel to the robot's x-direction (forward) and with the same x-offset.
+		// No slip is assumed (Instantaneous Center of Curvature (ICC) along the axis connecting the wheels).
+		// All computations are performed for ROS frame conventions.
+		case 2:
```

Comment:
I think be better to move case content to other function, e.g. `update_odometry_diffdrive()`.